File: javascript/packages/highlighter/test/document.test.ts
import { describe, it, expect, beforeEach, afterEach } from "vitest"
import { themes } from "../src/themes.js"

import { Herb } from "@herb-tools/node-wasm"
import { SyntaxRenderer } from "../src/syntax-renderer.js"

import type { StyledRun, StyleRole } from "../src/document.js"

const FIXTURE = "<div class=\"a\">\n  <!-- note -->\n  <% if x %>ok<%= y %>\n</div>"

const roleLabel = (role: StyleRole): string => {
  if (role.kind === "Token") {
    return `Token(${role.tokenType})`
  }

  return role.kind
}

const escapeText = (text: string): string =>
  text.replace(/\\/g, "\\\\").replace(/\n/g, "\\n")

const serialize = (runs: StyledRun[]): string =>
  runs.map((run) => `${roleLabel(run.role)}|${escapeText(run.text)}`).join("\n")

describe("Document", () => {
  let renderer: SyntaxRenderer
  let originalNoColor: string | undefined

  beforeEach(async () => {
    originalNoColor = process.env.NO_COLOR
    delete process.env.NO_COLOR

    renderer = new SyntaxRenderer(themes.onedark, Herb)
    await renderer.initialize()
  })

  afterEach(() => {
    if (originalNoColor === undefined) {
      delete process.env.NO_COLOR
    } else {
      process.env.NO_COLOR = originalNoColor
    }
  })

  it("serializes styled runs for a mixed fixture", () => {
    const runs = renderer.highlightRuns(FIXTURE)

    expect(serialize(runs)).toMatchSnapshot()
  })

  it("lex failure produces a single plain run", async () => {
    const errorHerb = {
      load: async () => {},
      lex: () => ({ errors: ["error"], value: [] }),
      isLoaded: true,
    }

    const errorRenderer = new SyntaxRenderer(themes.onedark, errorHerb as any)
    await errorRenderer.initialize()

    const content = "<invalid>"
    const runs = errorRenderer.highlightRuns(content)

    expect(runs).toEqual([{ text: content, role: { kind: "Plain" } }])
  })

  it("empty content produces no runs", () => {
    const runs = renderer.highlightRuns("")

    expect(runs).toEqual([])
  })
})

File: rust/herb-highlighter/tests/document_test.rs
mod common;

use std::sync::Arc;

use herb_highlighter::document::{StyleRole, StyledRun};
use herb_highlighter::herb_backend::Herb;
use herb_highlighter::syntax_renderer::SyntaxRenderer;
use herb_highlighter::themes::{get_theme, Theme};

use common::{with_color, StubBackend};

fn renderer(theme: Theme) -> SyntaxRenderer {
  SyntaxRenderer::with_backend(get_theme(theme).clone(), Arc::new(Herb))
}

fn role_label(role: &StyleRole) -> String {
  match role {
    StyleRole::Plain => "Plain".to_string(),
    StyleRole::Token(token_type) => format!("Token({token_type})"),
    StyleRole::RubyKeyword => "RubyKeyword".to_string(),
    StyleRole::TagName => "TagName".to_string(),
    StyleRole::AttributeName => "AttributeName".to_string(),
    StyleRole::AttributeValue => "AttributeValue".to_string(),
    StyleRole::CommentInterior => "CommentInterior".to_string(),
  }
}

fn escape_text(text: &str) -> String {
  text.replace('\\', "\\\\").replace('\n', "\\n")
}

fn serialize_runs(runs: &[StyledRun]) -> String {
  runs
    .iter()
    .map(|run| format!("{}|{}", role_label(&run.role), escape_text(&run.text)))
    .collect::<Vec<_>>()
    .join("\n")
}

#[test]
fn serializes_styled_runs_for_a_mixed_fixture() {
  with_color();

  let content = "<div class=\"a\">\n  <!-- note -->\n  <% if x %>ok<%= y %>\n</div>";

  insta::assert_snapshot!(serialize_runs(&renderer(Theme::OneDark).highlight_runs(content)));
}

#[test]
fn lex_failure_produces_a_single_plain_run() {
  with_color();

  let failing_renderer = SyntaxRenderer::with_backend(get_theme(Theme::OneDark).clone(), StubBackend::failing());

  let content = "<invalid>";

  assert_eq!(
    failing_renderer.highlight_runs(content),
    vec![StyledRun {
      text: content.to_string(),
      role: StyleRole::Plain,
    }]
  );
}

#[test]
fn empty_content_produces_no_runs() {
  with_color();

  assert_eq!(renderer(Theme::OneDark).highlight_runs(""), Vec::new());
}
